src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -817,6 +818,26 @@ class GrowableArrayCHeap : public GrowableArrayWithAllocator<E, GrowableArrayCHe
     this->clear_and_deallocate();
   }
 
+  GrowableArrayCHeap(GrowableArrayCHeap<E,F>& other)
+  : GrowableArrayWithAllocator<E, GrowableArrayCHeap<E, F>>(allocate(0, F), 0) {
+    this->grow(other._len);
+    for (int i = 0; i < other._len; i++) {
+      this->_data[i] = other._data[i];
+    }
+    this->_len = other._len;
+  }
+
+  GrowableArrayCHeap<E,F>& operator=(const GrowableArrayCHeap<E,F>& other) {
+    assert(&other != this, "must be different");
+    this->clear_and_deallocate();
+    this->grow(other._len);
+    for (int i = 0; i < other._len; i++) {
+      this->_data[i] = other._data[i];
```

Comment:
Does this imply that anything you put in a `GrowableArrayCHeap` must itself also be copyable now?

Reply:
Hi,

No, if you never call the assignment operator then it won't be generated and so does not imply that. If you have something `NONCOPYABLE` and you do call the assignment operator then compilation will fail.

However, just to be clear: It is already the case that everything we put into a `GrowableArray`, CHeap or not, must be copyable. This is a direct consequence of the fact that we do not allow for move semantics in HotSpot.

Reply:
Hmmm I'd guess I'd normally store pointers then, because I generally may not want to either move or copy the object that is accessible through the container. But thanks for clarifying.